### PR TITLE
Generate sample data deterministically from packet UUID.

### PR DESCRIPTION
**Pull request by BranLwyd**
The UUID is itself random, so sample data is still random; the upside of
this is that we can now determine a given sample ingestion packet's data
without reading & decrypting the data shares. This will be handy for a
validator which wants to make sure the sums we are emitting match up
with the ingestion packets included in those sums.

[Note that since even "random" v4 UUIDs contain a few deterministic
bytes, part of the data will be deterministic; I think that's OK.]

### CODE REVIEW

**Comment by BranLwyd** [comment]
I'm doing this to address #226, at least as far as validating that the sums we output are equal to the sum of the ingestion packets that go into them. My basic implementation strategy is something like:

1. Generate sample data in a way that allows the data to be determined from the packet itself [this PR].
2. Write a validator job that:
  a. Reads sum parts, computes actual sum.
  b. Reads all validation batches included in the sum, based on the validation batch IDs.
  c. Computes expected sum by computing from ingestion UUIDs in validation batches.
  d. Yells if the expected & actual sums don't match up.

The biggest failure case this would miss would be something of the form "we somehow completely ignored an ingestion packet" -- I haven't thought that part through too hard, but we could potentially later extend the validator job to also look for ingestion packets that "should" be included in the validation batches but aren't, and yell about them too.

Note this also ignores the various "bug-simulation" features of the sample generator (short packet, dropped packets) -- since the sample generator job doesn't use them, that's OK for now.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]
This is quite clever, but I'm not sure it'll quite work. You are deriving inputs from _packet_ UUIDs, but the portal server/collector/sum verifier you'll eventually write will have the list of _batch_ UUIDs, so you won't be able to rederive the inputs on that end. However, I think you could instead make the derivation a deterministic function of `(batch UUID, packet index)`. Then, on the sum verifier end, if you get an aggregation with 100 contributions over 10 batches, you can assume that each batch contained 10 inputs and re-derive the inputs.

**Comment by BranLwyd** [comment]
Ugh, you're correct of course. The sequence of values is now deterministic based on `batch_id` (and `dimension`, for the length). I also pulled this logic out to a separate function, since I'll need to be able to call it from elsewhere soon anyway.

File: facilitator/src/sample.rs [hunk 1]
```diff
@@ -339,6 +335,43 @@ impl<'a> SampleGenerator<'a> {
     }
 }
 
+/// Returns an iterator over sample datas; each will be a vector of length
```
Comment (tgeoghegan) [inline facilitator/src/sample.rs:338]:
```suggestion
/// Returns an iterator over sample data; each datum will be a vector of length
```

**Review by tgeoghegan — APPROVED** [review]
Rather elegant use of an iterator! I'd say that we should use an HMAC instead of the bespoke SHA256 construction but the output isn't security critical.

**Comment by BranLwyd** [comment]
> Rather elegant use of an iterator! I'd say that we should use an HMAC instead of the bespoke SHA256 construction but the output isn't security critical.

BTW -- this doesn't matter much/at all, but I'd argue that the existing hash-based construction is preferable to a MAC-based construction. I'm trying to implement a collision-resistant function over (batch_uuid, index). I presume you're suggesting `MAC(K = batch_uuid, M = index)` [or possibly flipping key/message, it's arbitrary]. I prefer the current implementation of `H((batch_uuid, index))`, with the tuple serialized as a concatenation of the bytes of its parts (which works since both parts are of constant size). Both constructions produce a collision-resistant value based on (batch_uuid, index); the advantage of the hash-based construction over the MAC-based construction is that the hash-based construction is slightly simpler.

**Pull request merged**